20240912-R-scheduler-auto-scale.md

```diff
@@ -0,0 +1,110 @@
+Schedulers and secondly enhancing the Schedulers such that we optionally run etcd instances in the Schedulers under
+certain circumstances that make sense. When scaling beyond 5 etcd instances, there are little performance improvements, 
+so it makes sense for Schedulers to become stateless at this point. For example, we could have 20 Schedulers, but only
+5 of them would contain the embedded etcd.
+
+## Related Items
+
+### Related issues
+
+[Tracking Issue: Scheduler Stable](https://github.com/dapr/dapr/issues/7796) &&
+[Auto Scale Scheduler Issue](https://github.com/dapr/dapr/issues/7912).
+
+## Expectations and alternatives
+
+There will be 3 changes needed for this feature to be complete. 
+
+1. **New control loop** for the [go-etcd-cron library](https://github.com/diagridio/go-etcd-cron), such that the leadership table is able to be dynamic. This 
+means we need to watch the leadership table to observe a new Scheduler starting up or going down, stopping all Schedulers 
+job triggers for that moment until the leadership table is updated and synchronized. We will need to stop the job queue 
+and rebuild it again by reshuffling all the jobs to evenly split up the job ownership amongst Schedulers. Refactoring is
+needed here such that the queue logic can be stopped and restarted. The leadership table will be updated to 
+include the host address as the data such that the key in the table remains the id, but the value becomes the host address.
+
+- An optimization we can do after the initial work is done, is to keep all jobs in memory. Having a local cache 
+would optimize building the queue faster, since we are using etcd this should be feasible. 
+
+2. **A new WatchHosts RPC** is added to `scheduler.proto`, allowing the dapr sidecar to connect to any available Scheduler 
+instance via a ClusterIP service, rather than needing to manually track and connect to each instance as in the current 
+headless service model. This `WatchHosts` RPC dynamically provides the current set of active Scheduler instances to the 
+sidecar, keeping it updated as Schedulers scale up or down.
+
+- For **backward compatibility**, we will check to determine whether the `WatchHosts` RPC is available or not. If it 
+exists, Dapr will use it to manage Scheduler connections dynamically. If not, Dapr will continue using the existing 
+static connection logic, maintaining compatibility with earlier versions by sending the full list back of available 
+Schedulers.
+
+```protobuf
+service Scheduler {
+  ...
+  rpc WatchHosts(WatchHostsRequest) returns (stream WatchHostsResponse) {}
+}
+
+message HostPort {
+  string host = 1;
+  uint32 port = 2;
+}
+
+message WatchHostsResponse {
+  repeated HostPort hosts = 1;
+}
+
+message WatchHostsRequest {}
+
+```
+
+3. **1/3/5 Schedulers with respective embedded etcd instances, beyond 5 instances run Schedulers stateless**. We want to
```

Comment:
I don't think this is feasible. The problem here is that when you scale down, from 5 to 3 instances, for example. Suddenly, the etcd cluster "thinks" that it is down 2 instances of the 5 and does not know that the other 2 were decommissioned. Instead, I vote for the cluster to have a fixed number of instances that run with etcd: 1 or 3 and that is the minimum number of replicas the user can have in the statefulset.

Reply:
Ive update the proposal to follow your suggestion of 1 or 3 etcd instances and not 5, where 3 is now our practical limit